syvita/daoos
# bootstrapping and main navigation

### Summary of Changes

- Boilerplate code for granular control over rendering (more on this later). This affects how future components will be rendered as well.
- Implementation of main navigation for desktop and responsive. Our nav system is custom and 100% from scratch since Material UI does not provide a pre-built nav bar that conforms strictly to our design specs.

### Please Note!

- The responsive breakpoints that are set are completely arbitrary and are subject to change and very likely will. I'd like to wait to see what it looks like when I start adding content before committing to specific breakpoints.
- There is a known issue with scroll "bouncing" on mobile browsing and safari desktop. What I mean by this is that the Nav Bar shouldn't bounce and reveal whitespace on an overscroll because this looks weird (you will see what I mean). There is an easy css fix for this on chrome desktop via `overscroll-behavior-y: none;` so the ideal behavior should be observed on chrome desktop. I will be making this into a separate/new issue since css solutions did not work for me without unwanted tradeoffs for mobile and safari. The fix for these platforms will likely involve client-side javascript. **Update: this should be fixed on all platforms except for desktop safari**
- I noticed the [desktop](https://www.figma.com/proto/F8qFmJv4Vq90zfR9QHnaGw/Dao-UI?node-id=92%3A1171&scaling=min-zoom) design does not have the invite and rules elements in the nav bar like the [mobile](https://www.figma.com/proto/qgJqserxx5gP4vWiCj9PCX/DAO-UI?node-id=33%3A2498&scaling=min-zoom) design does. I added these to the desktop nav bar anyways, since I believe that was the intention. **Correct me if i'm wrong and I will remove these elements from desktop**.
- In the mobile design the opacity of outside content is reduced when the nav takes focus. This was intentionally not implemented yet since we don't have outside content. Will add later.
- `rem` units are used for navbar links. What this means is that the text of these links will resize up or down depending on your viewport's zoom level. Therefore, depending on your viewport, the scaling of this text will differ from what you see in the design files. **This is intentional but let me know if you would otherwise prefer a constant sized nav text across all platforms.**
- Typescript checks are failing on this PR because the packages aren't installed at the root level where the github action checks are run. This shouldn't have to be the case, so I'm going to create an issue for resolving this. The typescript checks are actually passing in the `packages/ui` directory.
- `!important` flags are used in a lot of css elements to override Material UI styling (I don't like using their `useStyles` system for styling). I will have a new PR to change the styling order such that css modules can be used without `!important` overrides.
- "LOGO" text will obviously be replaced with an actual logo in a future PR :)

### Rendering Approach
We are using NextJS which is preconfigured and biased for SSR (server side rendering) and SSG (static site generation). However, our site has quite a different responsive UX from the desktop UX. The challenge this presents is that you don't know at request time on the server what experience to serve the client (i.e what breakpoints it needs) so strict conditional rendering can't be performed server-side the same way it is usually performed client-side in the React ecosystem. Here are the approaches that were considered:

- There is a `user-agent` request header that will _sometimes_ contain "mobile" if a mobile device is making the request. This could be parsed server side and used to deliver the appropriate data. However, this isn't always an accurate approach and there's no way of knowing exactly what breakpoints would need to be rendered for the device.
- pure css media queries. The problem with this approach is that it doesn't make use of conditional rendering. All of the elements outside of the breakpoints would still have to be mounted to the DOM with `display: none` which is non-performant and hacky.

I decided on using the [@artsy/fresnel](https://github.com/artsy/fresnel) library. fresnel renders all breakpoints on the server and delivers them to the client which will then take the one it currently needs and conditionally render the rest. The main downside is a larger initial page size but this additional overhead gets removed when the React application boots. This is worth the tradeoff in my opinion since you get all of the benefits of SSG, css media queries, and conditional rendering.

### How to run

~~1. install packages via `npm i` in `packages/ui` directory.
2. `npm run dev` from the same directory to launch the app.
3. App will start on `http://localhost:3000`. You can run from your browser or any device on the same wifi network (this will be important for mobile testing).~~
Just use the public cloudflare preview url below

### Test Criteria

- Test on both desktop and mobile.
- Test landscape and portrait.
- Ensure the implementation conforms to the design.
- Provide suggestions and/or feedback.
- Ensure behaviors are as expected.
- These are rough guidelines, please test and try to break with your own methods.

Report issues/feedback/suggestions as comments on the PR. Pease check the "Please Note!" section above and PR comments below first to see if the issue you're about to report has already been acknowledged.

This will close #55
